### PR TITLE
[move-stdlib] Add basic ASCII string support to the Move Stdlib

### PR DESCRIPTION
**Pull request by tzakian**
This PR adds a basic `ASCIIString` newtype wrapper around backing `payload` bytes that verifies that all characters are valid ASCII when the string is created.

This PR adds both unit tests and specifications for the module. One thing that is worth mentioning is I've deliberately tried to keep the function-level specs as minimal as possible and instead did the minimal needed to specify abort conditions and ensure the invariant on `ASCIIString` holds. The only real function-level specs are on the `all_characters_eye_visible` predicate in this module.

### CODE REVIEW

**Comment by xli** [comment]
Could we keep it simple, as it is ascii string, just support all valid chars. So that the document can be simple and people knows what need to do to handle error (if set the value failed due to invalid chars)

**Comment by xli** [comment]
> > I think in this case it would maybe make more sense to add an additional `Printable` module that encodes only printable strings. Clients could then determine off-chain by the type of the string if it is an ASCII or printable string.
> 
> My impression is that the folks who are most eager to have a string type want it exactly for the purpose of identifying data that should be printed as strings instead of vectors of bytes. `Printable` seems like the right thing for that. I'm not sure what to suggest, though, because we may also add UTF8 Unicode strings which may or may not be printable (depending if they contain newlines, etc.) I don't think we want all combinations of those.

The reason we're asking for string type, is not because we want to print them, it's more of recognize them, and communicate the type with developers. 
Communication is the main reason.
Thus defining a string type with clear message is better.
As we focusing on ASCII string, then it should be ASCII string without additional communication to tell what exactly it is.

**Comment by bob-wilson** [comment]
OK, it sounds like this PR will work for what @xli wants. The `all_characters_printable` function is totally optional. (I'm not sure if it's important to provide that function, but it certainly won't hurt anything.)

**Comment by tzakian** [comment]
Agreed that this sounds like this PR will work. I can remove the printable chars/strings functions if we want, but I think keeping them in the module for now is nice to have.

**Review by tnowacki — COMMENTED** [review]
This looks good to me! since it is adding something new to the stdlib, any thoughts on how to get consensus before landing?

**Comment by tzakian** [comment]
I'll make a post on the slack channel to see if there's anyone against landing this in its current form :) (if not, we can also discuss at the team meeting later today)

**Comment by vgao1996** [comment]
Looks good to me!

File: diem-move/diem-framework/experimental/releases/artifacts/current/build/DiemExperimental/sources/MultiToken.move [hunk 1]
```diff
@@ -84,7 +84,7 @@ module MultiToken {
         let owner_addr = GUID::id_creator_address(id);
         let tokens = &mut borrow_global_mut<TokenDataCollection<TokenType>>(owner_addr).tokens;
         let index_opt = index_of_token<TokenType>(tokens, id);
-        assert(Option::is_some(&index_opt), Errors::invalid_argument(EWRONG_TOKEN_ID));
+        assert!(Option::is_some(&index_opt), Errors::invalid_argument(EWRONG_TOKEN_ID));
```
Comment (tnowacki) [inline diem-move/diem-framework/experimental/releases/artifacts/current/build/DiemExperimental/sources/MultiToken.move:87]: why is this change here?

**Review by tnowacki — DISMISSED** [review]

File: diem-move/diem-framework/experimental/releases/artifacts/current/build/DiemExperimental/sources/MultiToken.move (hunk 1 above)
Comment (tzakian) [reply diem-move/diem-framework/experimental/releases/artifacts/current/build/DiemExperimental/sources/MultiToken.move:87]: The experimental release wasn't compiled in previous PRs -- this PR recompiles it, and it's updating these because of that. I'm going to be putting up a PR adding a test to make sure the experimental folder is kept up-to-date as well.

**Review by tnowacki — DISMISSED** [review]
